pkg/types/contract_reader_cache.go

```diff
@@ -0,0 +1,441 @@
+package types
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils"
+)
+
+var (
+	ErrNotExist        = errors.New("data does not exist for key")
+	ErrTypeMismatch    = errors.New("data type mismatch")
+	ErrInvalidStrategy = errors.New("invalid cache strategy")
+)
+
+const (
+	CachedContractReaderName = "CachedContractReader"
+
+	cleanInterval = time.Second
+	pollInterval  = time.Second
+)
+
+type TriggerAndCache struct {
+	Params any
+	Value  any
+}
+
+type PollAndCache struct {
+	TriggerAndCache
+	Interval time.Duration
+}
+
+type FallthroughOnStale struct {
+	Staleness time.Duration
+}
+
+type NoCache struct{}
+
+var _ ContractReader = &CachedContractReader{}
+var _ services.Service = &CachedContractReader{}
+
+type CachedContractReader struct {
+	UnimplementedContractReader
+
+	// injected dependencies
+	base ContractReader
+
+	// internal configuration properties
+	strategies map[string]any
```

Comment:
Why are strategies values type `any`?

Reply:
This allows multiple strategy types to be in the same key lookup. A specific read key has a cache strategy, but each could be configured differently.

Reply:
Unrestricted `any` doesn't appear to be necessary to achieve this, and makes the API very difficult to read. What about an interface type, just to contain the set of possible strategies?
```go
type CacheStrategy interface {
	cacheStrategy()
}
```